overlord/auth/auth.go

```diff
@@ -143,9 +144,21 @@ NextUser:
 	return nil, ErrInvalidAuth
 }
 
-// Authenticator returns MacaroonAuthenticator for current authenticated user represented by UserState
-func (us *UserState) Authenticator() *MacaroonAuthenticator {
-	return newMacaroonAuthenticator(us.StoreMacaroon, us.StoreDischarges)
+// Authenticator returns a store.Authenticator which uses the given
+// user's stored credentials, or nil if there are no credentials.
+func Authenticator(st *state.State, userID int) (store.Authenticator, error) {
+	var us *UserState
+	if userID != 0 {
+		var err error
+		us, err = User(st, userID)
```

Comment:
And then this is picking the user out of the request one more time, third since we entered the request, in the good case when we don't look for the auther multiple times.

Let's also talk about this please. I'm pretty sure we can do without, but I need a better view on your final idea.